### PR TITLE
[DSM] PEPPER-294 Fix for uploading kit with UUID and Fix for collaborator sample id counter

### PR DESCRIPTION
**Pull request by pegahtah**
## Context

This PR has fixes for 2 bugs:

1. When uploading a new kit for a participant that has some kits with pepper short id and some with legacy id, the new kit's collaborator sample id reflects the total previous number of kits 
2. When a kit is being uploaded using any id that is not a short id for studies that are Pepper studies (have a DSS side, so not Darwin) except for RGP, DSM shows an error that says the id that is used for upload is not a short id and therefore it is not valid

## Checklist

- [ ] I have labeled the type of changes involved using the `C-*` labels.
- [ ] I have assessed potential risks and labeled using the `R-*` labels.
- [ ] I have considered error handling and alerts, and added `L-*` labels as needed.
- [ ] I have considered security and privacy, and added `I-*` labels as needed
- [ ] I have analyzed my changes for stability, fault tolerance, graceful degradation, performance bottlenecks and written a brief summary in this PR.
- [ ] If applicable, I have discussed the analytics needs at both a platform and study level with Product and instrumented code accordingly.
- [ ] If applicable, my UI/UX changes have passed muster with Product/Design via an over-the-shoulder review, screenshots, etc.

_If unsure or need help with any of the above items, add the `help wanted` label. For items that starts with `If applicable`, if it is not applicable, check it off and add `n/a` in front._

## FUD Score

_Overall, how are you feeling about these changes?_

- [ ] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [ ] They are user-visible in dev as a regular user journey and require no additional instructions.
- [ ] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [ ] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [ ] I have added regression test labels to the jira ticket
- [ ] I have added verification steps to the jira ticket
- [ ] I have written automated positive tests
- [ ] I have written automated negative tests
- [ ] I have written zero automated tests but have poked around locally to verify proper functionality
- [ ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [ ] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

### CODE REVIEW

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java [hunk 1]
```diff
@@ -1875,23 +1920,6 @@ public static String getCollaboratorSampleId(int kitTypeId, String participantCo
         }
     }
 
-    public static String getParticipantIdFromLegacyKit(DDPInstance ddpInstance, String shortId, String collaboratorParticipantId) {
```
Comment (pegahtah) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java:1878]: Not used

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java [hunk 2]
```diff
@@ -1245,13 +1248,26 @@ public static String generateBspParticipantID(String collaboratorIdPrefix, Strin
...
                 collaboratorSampleId += "_" + type;
             }
+            String participantId = collaboratorParticipantId.split("_")[1];
             int counter = getKitCounter(conn, collaboratorSampleId, kitTypeId);
+
+            if  (ddpInstance.isMigratedDDP()) {
+                if (!ParticipantUtil.isHruid(participantId)) {
+                    //assuming participantId is a legacy short id
+                    counter += getPepperKitCounter(conn, collaboratorSampleId, participantId, kitTypeId,
+                            ddpInstance.getParticipantIndexES());
```
Comment (pegahtah) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java:1271]: For Migrated studies, if the legacy id is given, count the number of kits that were created with this participant's pepper short id and consider that number when creating the collaborator sample id for this kit. 

For example, if participant PABCDE has legacy short id 1, and has a saliva kit 0001_SALIVA_1 and PABCDE_SALIVA_2, the next kit should be 0001_SALIVA_3, and not 0001_SALIVA_2

Comment (zyme) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/KitRequestShipping.java:1271]: This comment is gold!  Please put it into source as a comment.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/statics/DBConstants.java [hunk 3]
```diff
@@ -2,6 +2,8 @@
 
 public class DBConstants {
 
+    public static final String RGP = "RGP";
```
Comment (zyme) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/statics/DBConstants.java:5]: How about moving the logic for "isRGP()" into DDPInstance?  A bit easier to find for re-use, and more general purpose than `isRgpUpload()`, which implies it's only suitable for upload-related checks.

File: pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/KitUploadRoute.java [hunk 4]
```diff
@@ -528,6 +537,10 @@ private String userExistsInRealm(DDPInstance ddpInstanceByRealm, Map<String, Str
         return checkKitUploadNameMatchesToEsName(participantFirstNameFromDoc, participantLastNameFromDoc, participantByShortId);
     }
 
+    private boolean isRgpUpload(DDPInstance ddpInstanceByRealm) {
```
Comment (zyme) [inline pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/KitUploadRoute.java]: Do we have PW coverage for RGP uploads?  If not, do we need a unit test to cover this scenario?

Comment (pegahtah) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/KitUploadRoute.java]: I believe we have them for rgp kits, which gives me relief

Comment (Rinchan20) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/KitUploadRoute.java]: We have PW coverage for the following RGP uploads:

- blood and rna for the proband/main study participant
- blood and rna for the proband's family member

Screenshot 2024-06-18 at 9 57 40 AM

Comment (Rinchan20) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/KitUploadRoute.java]: Just ran it for DSM Test - this is probably a better picture:
Screenshot 2024-06-18 at 10 01 35 AM

Comment (pegahtah) [reply pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/KitUploadRoute.java]: Thank you so much Kiara @Rinchan20

**Review by zyme — APPROVED** [review]

**Pull request merged**